ouroboros-consensus/test/storage-test/Test/Ouroboros/Storage/ChainDB/Model.hs

```diff
@@ -446,7 +463,49 @@ addBlock cfg blk m = Model {
 
     -- Note that this includes the currently selected chain, but that does not
     -- influence chain selection via 'selectChain'.
-    consideredCandidates = filter (extendsImmutableChain . fst) candidates
+    consideredCandidates =
+      candidates
+        & filter (extendsImmutableChain . fst)
+        & mapMaybe (bimapM trimToLoE pure)
+
+    currentChain' = currentChain m
+
+    -- If the LoE fragment does not intersect with the current selection, then
+    -- we use the empty fragment anchored at the immutable tip instead.
+    loeFragment' =
+      loeFragment m <&> \loeFragment'' ->
+        case Fragment.intersect (Chain.toAnchoredFragment currentChain') loeFragment'' of
+          Just _ -> loeFragment''
+          Nothing -> Fragment.Empty $ Chain.headAnchor $ immutableChain secParam m
+
+    -- REVIEW: This might lead to 'consideredCandidates' containing duplicates;
+    -- is this a problem?
+    trimToLoE :: Chain blk -> Maybe (Chain blk)
+    trimToLoE chain =
```

Comment:
An alternative implementation that doesn't have any `error` calls and is more different from the "actual" implementation:
```haskell
    trimToLoE :: Chain blk -> Chain blk
    trimToLoE candidate =
      case loeChain of
        LoEDisabled          -> candidate
        LoEEnabled loeChain' ->
          Chain.fromOldestFirst $ go (Chain.toOldestFirst candidate) loePoints
          where
            loePoints = blockPoint <$> Chain.toOldestFirst loeChain'
      where
        SecurityParam k = secParam

        go :: [blk] -> [Point blk] -> [blk]
        -- The LoE chain is an extension of the candidate, return the candidate.
        go []           _loePoints       = []
        -- The candidate is an extension of the LoE chain, return at most the
        -- next k blocks on the candidate.
        go blks         []               = take (fromIntegral k) blks
        go (blk : blks) (pt : loePoints)
          -- The candidate and the LoE chain agree on the next point, continue
          -- recursively.
          | blockPoint blk == pt         = blk : go blks loePoints
          -- The candidate forks off from the LoE chain; stop here.
          | otherwise                    = []
```
together with
```haskell
    -- If the LoE fragment does not intersect with the current selection, then
    -- we use the immutable chain instead.
    loeChain =
      loeFragment m <&> \loeFragment' ->
        case Fragment.cross (Chain.toAnchoredFragment currentChain') loeFragment' of
          Just (_, loeChain') ->
            fromMaybe (error "currentChain' is not anchored at Origin") $
              Chain.fromAnchoredFragment loeChain'
          Nothing        -> immutableChain'

    immutableChain' = immutableChain secParam m

    immutableChainHashes =
        map blockHash
      . Chain.toOldestFirst
      $ immutableChain'
```
It seems to pass the test. WDYT?

Reply:
It looks OK to me! I agree that it is nice to separate from the actual implementation (even though this modelled implementation existed before the other!) and this looks good to me.

Reply:
Done in 3e6a7bb818ffc6738facce35ba0cc648b8e171e8.